### ISSUE TITLE
Allow configuration of pundit_user (aka custom UserContext)

### ISSUE THREAD

**Issue by thibaudgg** [venuu/jsonapi-authorization#12]
Like specified on [Pundit README](https://github.com/elabs/pundit#additional-context) it would be nice to add a configuration option to define a custom `pundit_user`. 
At the moment this hardcoded as `context[:user]` in this gem. This is not a problem in [`DefaultPunditAuthorizer`](https://github.com/venuu/jsonapi-authorization/blob/master/lib/jsonapi/authorization/default_pundit_authorizer.rb#L22) as its main purpose is to be edited/replaced but I think this is an issue in the `PunditScopedResource` class ([here](https://github.com/venuu/jsonapi-authorization/blob/master/lib/jsonapi/authorization/pundit_scoped_resource.rb#L10) and [here](https://github.com/venuu/jsonapi-authorization/blob/master/lib/jsonapi/authorization/pundit_scoped_resource.rb#L22)) as it it force us to overwrite the class.

WDYT?

**Comment by lime** [comment]
Very true, I had forgotten that we accessed `context[:user]` from there as well.

We need this on both the class and instance level of the resource, as well as in the default authorizer. What about a configurable proc that takes the context?
```
JSONAPI::Authorization.configure do |config|
  config.user = -> (context) { context[:user] }
end
```

**Comment by lime** [comment]
Note that most users should be fine without this, since they can already define the entire `context` any way they like.
```
def context
  { user: CustomUserThingy.new(user: current_user, foo: bar) }
end
```

I'm still in favour of making this configurable. Someone might already use the `:user` key for something else, or just prefer to name it differently. Most importantly though, it would consolidate the `context` access into one place.

**Comment by thibaudgg** [comment]
I know that you don't want to be too much bind to Pundit in particular but I've the feeling that `config.user` is too vague here, I think `config.pundit_user` would be more clear. Maybe `authorization_user` ?

``` ruby
JSONAPI::Authorization.configure do |config|
  config.authorization_user = -> (context) { context[:user] }
end
```

**Comment by valscion** [comment]
Naming is the most difficult thing :smile: `authorization_user` seems like a good name to me, what do you think @lime?

**Comment by lime** [comment]
Yep, I deliberately left it vague as I thought it might need some discussion. ;)

`authorization_user` is ok by me. In fact, so is `pundit_user`; we are after all only using it in our Pundit tie-ins.

An even more general term would be something like `authorizee`, but that's a bit of a mouthful. :stuck_out_tongue:

**Comment by thibaudgg** [comment]
Ok let's go with `pundit_user` then, I think it's clearer.

**Comment by valscion** [comment]
If someone wants to create a PR for this, we're open on accepting it :)

This is what we'd like to have:

```ruby
JSONAPI::Authorization.configure do |config|
  config.pundit_user = -> (context) { ...fetch user from context somehow... }
end
```

It should be an optional configuration and would default to fetching the user object from `context[:user]`.

There are three places where we fetch the user object, and they should be all changed to use the configuration:

1. https://github.com/venuu/jsonapi-authorization/blob/v0.6.0/lib/jsonapi/authorization/pundit_scoped_resource.rb#L10
2. https://github.com/venuu/jsonapi-authorization/blob/v0.6.0/lib/jsonapi/authorization/pundit_scoped_resource.rb#L15
3. https://github.com/venuu/jsonapi-authorization/blob/v0.6.0/lib/jsonapi/authorization/default_pundit_authorizer.rb#L22

**Comment by coryodaniel** [comment]
What do you think about pushing the entire context into the policy. Then someone could call the key whatever they want without requiring a configuration and it would provide more flexibility in an instance where you may want to do change permissions based on an api version...

```ruby
class CoolPolicy
  attr_reader :context, :user, :record

  def initialize(context, record)
    @context = context
    @user = context[:current_user]
    @record = record
  end

  def show?
    if context[:api_version] == 1
      Api::V1::CoolPolicy.new(user,record).show?
    else
      Api::V2::CoolPolicy.new(context,record).show?
    end
  end
end
```

**Comment by coryodaniel** [comment]
Added this feature -> https://github.com/venuu/jsonapi-authorization/pull/32

**Comment by valscion** [comment]
This is implemented in v0.8.1 thanks to @coryodaniel 💖

**Comment by thibaudgg** [comment]
Thanks @coryodaniel! :clap:
